lib/AST/Type.cpp

```diff
@@ -1137,6 +1137,79 @@ getCanonicalParams(AnyFunctionType *funcType,
   }
 }
 
+StringRef TypeBase::getKindName(TypeKind K) {
+#define ENTRY(Kind, String)                                                    \
+  case TypeKind::Kind:                                                         \
+    return String
+  switch (K) {
+    ENTRY(Error, "error type");
+    ENTRY(Unresolved, "unresolved type");
+    ENTRY(Hole, "hole type");
+    ENTRY(TypeVariable, "type variable type");
+
+    ENTRY(BuiltinInteger, "builtin type");
+    ENTRY(BuiltinIntegerLiteral, "builtin type");
+    ENTRY(BuiltinFloat, "builtin type");
+    ENTRY(BuiltinRawPointer, "builtin type");
+    ENTRY(BuiltinNativeObject, "builtin type");
+    ENTRY(BuiltinUnsafeValueBuffer, "builtin type");
+    ENTRY(BuiltinVector, "builtin type");
+    ENTRY(BuiltinBridgeObject, "builtin type");
+
+    ENTRY(Tuple, "tuple type");
+
+    ENTRY(WeakStorage, "reference storage type");
+    ENTRY(UnownedStorage, "reference storage type");
+    ENTRY(UnmanagedStorage, "reference storage type");
+
+    ENTRY(Enum, "enum type");
+    ENTRY(Struct, "struct type");
+    ENTRY(Class, "class type");
+    ENTRY(Protocol, "protocol type");
+
+    ENTRY(BoundGenericClass, "class type");
+    ENTRY(BoundGenericEnum, "enum type");
+    ENTRY(BoundGenericStruct, "struct type");
+
+    ENTRY(UnboundGeneric, "generic type");
+
+    ENTRY(Metatype, "metatype type");
+    ENTRY(ExistentialMetatype, "metatype type");
```

Comment:
There *is* a difference here. Also, I worry about surfacing "metatype", even worse "existential metatype"

Reply:
Yeah. I was just trying to “group” them. I think the those terms exist in the documentation IIRC, so it’s not too bad if it shows up (plus people have been talking about metatype extensions on forums...). 

Are there other concerns about showing it in diagnostics?

Reply:
I think "metatype" is the right user-facing term...it's obviously a more advanced concept, but I'm not sure what we would replace the term with otherwise.

I'm inclined to use "protocol metatype" instead of "existential metatype" since we generally say "protocol as a type" instead of "existential type", but I'm not actually sure if that's better here.

Reply:
The only user-facing use of “existential metatype” I can find is the [documentation](https://developer.apple.com/documentation/swift/2885064-type) for `type(of:)`. Apart from that, it seems to be used in Mirror when printing description/summary.

Reply:
Sorry, I should clarify - I don't think we should use the term "existential metatype", but I have yet to think of a good alternative!

Reply:
No worries. I think your suggestion about using “protocol metatype” is better. The only alternative I can think of is simply calling it “metatype”/“metatype type” and not trying to distinguish between the different kinds, similar to generic and non-generic function types which I have referred to as “function types”.